// @yikes2000/prettier-plugin-merge-extras/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __esm = (fn, res) => function __init() {
  return fn && (res = (0, fn[__getOwnPropNames(fn)[0]])(fn = 0)), res;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// node_modules/.pnpm/diff@5.1.0/node_modules/diff/lib/index.mjs
function Diff() {
}
function buildValues(diff2, components, newString, oldString, useLongestToken) {
  var componentPos = 0, componentLen = components.length, newPos = 0, oldPos = 0;
  for (; componentPos < componentLen; componentPos++) {
    var component = components[componentPos];
    if (!component.removed) {
      if (!component.added && useLongestToken) {
        var value = newString.slice(newPos, newPos + component.count);
        value = value.map(function(value2, i) {
          var oldValue = oldString[oldPos + i];
          return oldValue.length > value2.length ? oldValue : value2;
        });
        component.value = diff2.join(value);
      } else {
        component.value = diff2.join(newString.slice(newPos, newPos + component.count));
      }
      newPos += component.count;
      if (!component.added) {
        oldPos += component.count;
      }
    } else {
      component.value = diff2.join(oldString.slice(oldPos, oldPos + component.count));
      oldPos += component.count;
      if (componentPos && components[componentPos - 1].added) {
        var tmp = components[componentPos - 1];
        components[componentPos - 1] = components[componentPos];
        components[componentPos] = tmp;
      }
    }
  }
  var lastComponent = components[componentLen - 1];
  if (componentLen > 1 && typeof lastComponent.value === "string" && (lastComponent.added || lastComponent.removed) && diff2.equals("", lastComponent.value)) {
    components[componentLen - 2].value += lastComponent.value;
    components.pop();
  }
  return components;
}
function clonePath(path) {
  return {
    newPos: path.newPos,
    components: path.components.slice(0)
  };
}
function diffLines(oldStr, newStr, callback) {
  return lineDiff.diff(oldStr, newStr, callback);
}
function _typeof(obj) {
  "@babel/helpers - typeof";
  if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
    _typeof = function(obj2) {
      return typeof obj2;
    };
  } else {
    _typeof = function(obj2) {
      return obj2 && typeof Symbol === "function" && obj2.constructor === Symbol && obj2 !== Symbol.prototype ? "symbol" : typeof obj2;
    };
  }
  return _typeof(obj);
}
function canonicalize(obj, stack, replacementStack, replacer, key) {
  stack = stack || [];
  replacementStack = replacementStack || [];
  if (replacer) {
    obj = replacer(key, obj);
  }
  var i;
  for (i = 0; i < stack.length; i += 1) {
    if (stack[i] === obj) {
      return replacementStack[i];
    }
  }
  var canonicalizedObj;
  if ("[object Array]" === objectPrototypeToString.call(obj)) {
    stack.push(obj);
    canonicalizedObj = new Array(obj.length);
    replacementStack.push(canonicalizedObj);
    for (i = 0; i < obj.length; i += 1) {
      canonicalizedObj[i] = canonicalize(obj[i], stack, replacementStack, replacer, key);
    }
    stack.pop();
    replacementStack.pop();
    return canonicalizedObj;
  }
  if (obj && obj.toJSON) {
    obj = obj.toJSON();
  }
  if (_typeof(obj) === "object" && obj !== null) {
    stack.push(obj);
    canonicalizedObj = {};
    replacementStack.push(canonicalizedObj);
    var sortedKeys = [], _key;
    for (_key in obj) {
      if (obj.hasOwnProperty(_key)) {
        sortedKeys.push(_key);
      }
    }
    sortedKeys.sort();
    for (i = 0; i < sortedKeys.length; i += 1) {
      _key = sortedKeys[i];
      canonicalizedObj[_key] = canonicalize(obj[_key], stack, replacementStack, replacer, _key);
    }
    stack.pop();
    replacementStack.pop();
  } else {
    canonicalizedObj = obj;
  }
  return canonicalizedObj;
}
var characterDiff, extendedWordChars, reWhitespace, wordDiff, lineDiff, sentenceDiff, cssDiff, objectPrototypeToString, jsonDiff, arrayDiff;
var init_lib = __esm({
  "node_modules/.pnpm/diff@5.1.0/node_modules/diff/lib/index.mjs"() {
    Diff.prototype = {
      diff: function diff(oldString, newString) {
        var options7 = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : {};
        var callback = options7.callback;
        if (typeof options7 === "function") {
          callback = options7;
          options7 = {};
        }
        this.options = options7;
        var self = this;
        function done(value) {
          if (callback) {
            setTimeout(function() {
              callback(void 0, value);
            }, 0);
            return true;
          } else {
            return value;
          }
        }
        oldString = this.castInput(oldString);
        newString = this.castInput(newString);
        oldString = this.removeEmpty(this.tokenize(oldString));
        newString = this.removeEmpty(this.tokenize(newString));
        var newLen = newString.length, oldLen = oldString.length;
        var editLength = 1;
        var maxEditLength = newLen + oldLen;
        if (options7.maxEditLength) {
          maxEditLength = Math.min(maxEditLength, options7.maxEditLength);
        }
        var bestPath = [{
          newPos: -1,
          components: []
        }];
        var oldPos = this.extractCommon(bestPath[0], newString, oldString, 0);
        if (bestPath[0].newPos + 1 >= newLen && oldPos + 1 >= oldLen) {
          return done([{
            value: this.join(newString),
            count: newString.length
          }]);
        }
        function execEditLength() {
          for (var diagonalPath = -1 * editLength; diagonalPath <= editLength; diagonalPath += 2) {
            var basePath = void 0;
            var addPath = bestPath[diagonalPath - 1], removePath = bestPath[diagonalPath + 1], _oldPos = (removePath ? removePath.newPos : 0) - diagonalPath;
            if (addPath) {
              bestPath[diagonalPath - 1] = void 0;
            }
            var canAdd = addPath && addPath.newPos + 1 < newLen, canRemove = removePath && 0 <= _oldPos && _oldPos < oldLen;
            if (!canAdd && !canRemove) {
              bestPath[diagonalPath] = void 0;
              continue;
            }
            if (!canAdd || canRemove && addPath.newPos < removePath.newPos) {
              basePath = clonePath(removePath);
              self.pushComponent(basePath.components, void 0, true);
            } else {
              basePath = addPath;
              basePath.newPos++;
              self.pushComponent(basePath.components, true, void 0);
            }
            _oldPos = self.extractCommon(basePath, newString, oldString, diagonalPath);
            if (basePath.newPos + 1 >= newLen && _oldPos + 1 >= oldLen) {
              return done(buildValues(self, basePath.components, newString, oldString, self.useLongestToken));
            } else {
              bestPath[diagonalPath] = basePath;
            }
          }
          editLength++;
        }
        if (callback) {
          (function exec() {
            setTimeout(function() {
              if (editLength > maxEditLength) {
                return callback();
              }
              if (!execEditLength()) {
                exec();
              }
            }, 0);
          })();
        } else {
          while (editLength <= maxEditLength) {
            var ret = execEditLength();
            if (ret) {
              return ret;
            }
          }
        }
      },
      pushComponent: function pushComponent(components, added, removed) {
        var last = components[components.length - 1];
        if (last && last.added === added && last.removed === removed) {
          components[components.length - 1] = {
            count: last.count + 1,
            added,
            removed
          };
        } else {
          components.push({
            count: 1,
            added,
            removed
          });
        }
      },
      extractCommon: function extractCommon(basePath, newString, oldString, diagonalPath) {
        var newLen = newString.length, oldLen = oldString.length, newPos = basePath.newPos, oldPos = newPos - diagonalPath, commonCount = 0;
        while (newPos + 1 < newLen && oldPos + 1 < oldLen && this.equals(newString[newPos + 1], oldString[oldPos + 1])) {
          newPos++;
          oldPos++;
          commonCount++;
        }
        if (commonCount) {
          basePath.components.push({
            count: commonCount
          });
        }
        basePath.newPos = newPos;
        return oldPos;
      },
      equals: function equals(left, right) {
        if (this.options.comparator) {
          return this.options.comparator(left, right);
        } else {
          return left === right || this.options.ignoreCase && left.toLowerCase() === right.toLowerCase();
        }
      },
      removeEmpty: function removeEmpty(array) {
        var ret = [];
        for (var i = 0; i < array.length; i++) {
          if (array[i]) {
            ret.push(array[i]);
          }
        }
        return ret;
      },
      castInput: function castInput(value) {
        return value;
      },
      tokenize: function tokenize(value) {
        return value.split("");
      },
      join: function join(chars) {
        return chars.join("");
      }
    };
    characterDiff = new Diff();
    extendedWordChars = /^[A-Za-z\xC0-\u02C6\u02C8-\u02D7\u02DE-\u02FF\u1E00-\u1EFF]+$/;
    reWhitespace = /\S/;
    wordDiff = new Diff();
    wordDiff.equals = function(left, right) {
      if (this.options.ignoreCase) {
        left = left.toLowerCase();
        right = right.toLowerCase();
      }
      return left === right || this.options.ignoreWhitespace && !reWhitespace.test(left) && !reWhitespace.test(right);
    };
    wordDiff.tokenize = function(value) {
      var tokens = value.split(/([^\S\r\n]+|[()[\]{}'"\r\n]|\b)/);
      for (var i = 0; i < tokens.length - 1; i++) {
        if (!tokens[i + 1] && tokens[i + 2] && extendedWordChars.test(tokens[i]) && extendedWordChars.test(tokens[i + 2])) {
          tokens[i] += tokens[i + 2];
          tokens.splice(i + 1, 2);
          i--;
        }
      }
      return tokens;
    };
    lineDiff = new Diff();
    lineDiff.tokenize = function(value) {
      var retLines = [], linesAndNewlines = value.split(/(\n|\r\n)/);
      if (!linesAndNewlines[linesAndNewlines.length - 1]) {
        linesAndNewlines.pop();
      }
      for (var i = 0; i < linesAndNewlines.length; i++) {
        var line = linesAndNewlines[i];
        if (i % 2 && !this.options.newlineIsToken) {
          retLines[retLines.length - 1] += line;
        } else {
          if (this.options.ignoreWhitespace) {
            line = line.trim();
          }
          retLines.push(line);
        }
      }
      return retLines;
    };
    sentenceDiff = new Diff();
    sentenceDiff.tokenize = function(value) {
      return value.split(/(\S.+?[.!?])(?=\s+|$)/);
    };
    cssDiff = new Diff();
    cssDiff.tokenize = function(value) {
      return value.split(/([{}:;,]|\s+)/);
    };
    objectPrototypeToString = Object.prototype.toString;
    jsonDiff = new Diff();
    jsonDiff.useLongestToken = true;
    jsonDiff.tokenize = lineDiff.tokenize;
    jsonDiff.castInput = function(value) {
      var _this$options = this.options, undefinedReplacement = _this$options.undefinedReplacement, _this$options$stringi = _this$options.stringifyReplacer, stringifyReplacer = _this$options$stringi === void 0 ? function(k, v) {
        return typeof v === "undefined" ? undefinedReplacement : v;
      } : _this$options$stringi;
      return typeof value === "string" ? value : JSON.stringify(canonicalize(value, null, null, stringifyReplacer), stringifyReplacer, "  ");
    };
    jsonDiff.equals = function(left, right) {
      return Diff.prototype.equals.call(jsonDiff, left.replace(/,([\r\n])/g, "$1"), right.replace(/,([\r\n])/g, "$1"));
    };
    arrayDiff = new Diff();
    arrayDiff.tokenize = function(value) {
      return value.slice();
    };
    arrayDiff.join = arrayDiff.removeEmpty = function(value) {
      return value;
    };
  }
});

// src/packages/core-parts/extras/align-obj-props/options.ts
var options;
var init_options = __esm({
  "src/packages/core-parts/extras/align-obj-props/options.ts"() {
    options = {
      alignObjectProperties: {
        category: "Format",
        type: "string",
        default: "colon",
        since: "0.6.1",
        description: "Align object properties."
      },
      alignSingleProperty: {
        category: "Format",
        type: "boolean",
        default: true,
        since: "0.6.1",
        description: "Add space around colon even for objects of single property."
      }
    };
  }
});

// src/packages/core-parts/extras/align-obj-props/align-obj-props.ts
function alignObjectProperties(code, options7) {
  opts = options7;
  const segs = splitByCurly(code);
  const group = {};
  let ignoreAboveLvl = -1;
  for (let i = 0; i < segs.length; i++) {
    const seg = segs[i];
    const { indent } = seg.firstLine;
    Object.keys(group).map((indentLvl) => Number(indentLvl)).filter((indentLvl) => indentLvl > indent).forEach((indentLvl) => {
      if (ignoreAboveLvl === -1 || indentLvl <= ignoreAboveLvl) {
        alignObjectPropLevel(segs, group[indentLvl], indentLvl);
      }
      delete group[indentLvl];
    });
    group[indent] ||= [];
    group[indent].push(i);
    if (ignoreAboveLvl > -1) {
      if (indent <= ignoreAboveLvl) {
        ignoreAboveLvl = -1;
      }
    } else if (seg.beginIgnore) {
      ignoreAboveLvl = indent;
    }
  }
  return "".concat(...segs.map((x) => x.lines));
}
function getIndent(line) {
  return line.match(/^[ \t\n]*$/) ? 9999 : line.match(/^([ \t]*)/)[0].length;
}
function makeSegment(lines) {
  if (lines.match(/^[ \t\n]*$/)) {
    return {
      lines,
      firstLine: { indent: 9999 },
      lastLine: { text: lines, indent: 9999 }
    };
  }
  const firstLine = lines.match(/^(?:[ \t]*\n)*([ \t]*\S.*\n)/)[1];
  const lastLine = lines.match(/([ \t]*\S.*\n)(?:[ \t\n]*)$/)[1];
  const linesNoComment = lines.replace(/\s*\/\/[^\n]*(?=\n)/g, "");
  return {
    lines,
    firstLine: { text: firstLine, indent: getIndent(firstLine) },
    lastLine: { text: lastLine, indent: getIndent(lastLine) },
    startIsCloseCurly: RE_START_IS_CLOSE_CURLY.test(firstLine),
    startIsCloseSymbol: RE_START_IS_CLOSE_SYMBOL.test(firstLine),
    endIsOpenCurly: RE_END_IS_OPEN_CURLY.test(lastLine),
    beginIgnore: RE_END_IS_EOL_IGNORE.test(lastLine),
    endIsFuncOpenCurly: RE_END_IS_FUNC_OPEN_BRACE.test(linesNoComment) && !RE_END_IS_OBJ_OPEN_BRACE.test(linesNoComment)
  };
}
function splitByCurly(code) {
  const rawSegs = code.split(RE_CURLY_BRACE_SEGMENTATION).map((lines) => makeSegment(lines));
  const segs = [rawSegs.shift()];
  while (rawSegs.length) {
    const prev = segs[segs.length - 1];
    const next = rawSegs.shift();
    const openCloseBadIndent = prev.endIsOpenCurly && next.startIsCloseCurly && prev.lastLine.indent !== next.firstLine.indent;
    const openNotIncreaseIndent = prev.endIsOpenCurly && !next.startIsCloseCurly && prev.lastLine.indent >= next.firstLine.indent;
    const closeNotReduceIndent = next.startIsCloseSymbol && prev.firstLine.indent <= next.firstLine.indent;
    const prevSegInvalidIndent = prev.firstLine.indent > prev.lastLine.indent && !(prev.lastLine.text.match(/(?:^\s*[)}\]>:]|`,?)/) || prev.endIsOpenCurly);
    const mergeToPrev = openCloseBadIndent || openNotIncreaseIndent || closeNotReduceIndent || prevSegInvalidIndent;
    if (mergeToPrev) {
      segs[segs.length - 1] = makeSegment(prev.lines + next.lines);
    } else {
      segs.push(next);
    }
  }
  return segs;
}
function alignObjectPropLevel(segs, idx, indent) {
  if (idx[0] > 0) {
    const prevSeg = segs[idx[0] - 1];
    if (prevSeg.endIsFuncOpenCurly) {
      return;
    }
  }
  const RE_PROPERTY = getRegExpProperty(indent);
  const RE_COMMENT = new RegExp(`^[ ]{${indent}}//`);
  const RE_SPREAD_OR_PROP_SHORTHAND = new RegExp(`^[ ]{${indent}}(?:\\.\\.\\.)?\\w+,`);
  const isGroupContinuation = (line) => RE_PROPERTY.test(line) || RE_COMMENT.test(line) || RE_SPREAD_OR_PROP_SHORTHAND.test(line);
  const change = { count: 0 };
  const mySegs = idx.map((i) => segs[i]);
  mySegs.forEach((seg) => {
    const lines = seg.lines.split(/(?<=\n)/);
    const firstLine = lines.shift();
    let lastChunk = [firstLine];
    let chunks = [lastChunk];
    let isPrevChunkContinuing = isGroupContinuation(firstLine);
    while (lines.length) {
      const line = lines.shift();
      const isLineContinuingGroup = isGroupContinuation(line);
      if (!isPrevChunkContinuing && isLineContinuingGroup) {
        chunks.push(lastChunk = []);
      }
      lastChunk.push(line);
      isPrevChunkContinuing = isLineContinuingGroup;
    }
    chunks = chunks.map((chunk) => alignObjectPropGroup(chunk, indent, change)).map((chunk) => "".concat(...chunk));
    seg.chunks = chunks;
  });
  if (change.count > 0 && (change.count > 1 || opts.alignSingleProperty)) {
    mySegs.forEach((seg) => {
      seg.lines = "".concat(...seg.chunks);
    });
  }
}
function getRegExpProperty(indent) {
  return new RegExp(`^[ ]{${indent}}((?:[^'" :]+|(['"])[^\\2\\n]+?\\2|\\[[^\\]]+\\])\\??):`);
}
function alignObjectPropGroup(group, indent, change) {
  const RE_PROPERTY = getRegExpProperty(indent);
  let ticks = countBackTicks("".concat(...group));
  ticks = ticks % 2 ? 0 : ticks;
  const maxLen = group.filter((line) => RE_PROPERTY.test(line)).map((line) => line.match(RE_PROPERTY)[1].length).sort((a, b) => b - a)[0];
  group = group.map((line) => {
    const key = (line.match(RE_PROPERTY) || {})[1];
    const inBetweenTicks = ticks > 0 && ticks % 2;
    if (key && !inBetweenTicks) {
      const padLen = maxLen - key.length;
      line = line.replace(
        RE_PROPERTY,
        opts.alignObjPropBy === "colon" ? `${" ".repeat(indent) + key + " ".repeat(padLen + 1)}:` : `${" ".repeat(indent) + key}:${" ".repeat(padLen)}`
      );
      change.count++;
    }
    if (ticks > 0) {
      ticks -= countBackTicks(line);
    }
    return line;
  });
  return group;
}
function countBackTicks(code) {
  return code.split("").filter((c) => c === "`").length;
}
var opts, RE_CURLY_BRACE_SEGMENTATION, RE_START_IS_CLOSE_CURLY, RE_START_IS_CLOSE_SYMBOL, RE_END_IS_OPEN_CURLY, RE_END_IS_EOL_IGNORE, RE_END_IS_FUNC_OPEN_BRACE, RE_END_IS_OBJ_OPEN_BRACE;
var init_align_obj_props = __esm({
  "src/packages/core-parts/extras/align-obj-props/align-obj-props.ts"() {
    RE_CURLY_BRACE_SEGMENTATION = /(?<=\{(?:[ \t]*\/\/(?:[^\n]*\/\/)?)?\n)|(?<=\n)(?=[ \t]*[})\]])/s;
    RE_START_IS_CLOSE_CURLY = /^[ \t]*\}/;
    RE_START_IS_CLOSE_SYMBOL = /^[ \t]*[})\]]/;
    RE_END_IS_OPEN_CURLY = /\{(?:[ \t]*\/\/(?:[^\n]*\/\/)?)?\n[ \t\n]*$/;
    RE_END_IS_EOL_IGNORE = /\{(?:[ \t]*\/\/(?:[^\n]*\/\/)?)\n[ \t\n]*$/;
    RE_END_IS_FUNC_OPEN_BRACE = /[\w)>]\s+\{[ \t\n]*$/;
    RE_END_IS_OBJ_OPEN_BRACE = /\b(?:return|throw|interface \w+)\s+\{[^\n]*[ \t\n]*$/i;
  }
});

// src/packages/core-parts/extras/align-obj-props/postprocess.ts
function setOptions(options7) {
  alignObjPropBy = options7.alignObjectProperties ?? alignObjPropBy;
  if (!validOptions.includes(alignObjPropBy)) {
    [alignObjPropBy] = validOptions;
  }
  alignSingleProperty = !!options7.alignSingleProperty;
}
function postprocess(code, options7) {
  setOptions(options7);
  if (alignObjPropBy === "none") {
    return code;
  }
  return alignObjectProperties(code, { alignObjPropBy, alignSingleProperty });
}
var validOptions, alignObjPropBy, alignSingleProperty;
var init_postprocess = __esm({
  "src/packages/core-parts/extras/align-obj-props/postprocess.ts"() {
    init_align_obj_props();
    validOptions = ["colon", "value", "none"];
    alignObjPropBy = "colon";
    alignSingleProperty = false;
  }
});

// src/packages/core-parts/extras/align-obj-props/index.ts
var handleParsers, AlignObjectProperties;
var init_align_obj_props2 = __esm({
  "src/packages/core-parts/extras/align-obj-props/index.ts"() {
    init_options();
    init_postprocess();
    handleParsers = ["babel", "typescript", "angular"];
    AlignObjectProperties = {
      options,
      postprocess: (code, opts2) => handleParsers.includes(opts2.parserName) ? postprocess(code, opts2) : code
    };
  }
});

// src/packages/core-parts/extras/merge-imports/options.ts
var options2;
var init_options2 = __esm({
  "src/packages/core-parts/extras/merge-imports/options.ts"() {
    options2 = {
      mergeSimpleImports: {
        category: "Format",
        type: "boolean",
        default: true,
        since: "0.6.1",
        description: "Merge adjacent simple imports."
      }
    };
  }
});

// src/packages/core-parts/extras/merge-imports/postprocess.ts
function setOptions2(options7) {
  mergeImports = !!options7.mergeSimpleImports;
}
function postprocess2(code, options7) {
  setOptions2(options7);
  if (!mergeImports || code.match(/^\s*$/)) {
    return code;
  }
  return mergeSimpleImports(code, options7);
}
function mergeSimpleImports(code, options7) {
  const segs = code.split(RE_SPLIT_IMPORTS);
  const codeBody = segs[segs.length - 1].match(/^import\s+/m) ? "" : segs.pop();
  let importLines = "".concat(...segs);
  let changeCount = 0;
  while (importLines.match(RE_ADJ_IMPORTS)) {
    importLines = importLines.replace(RE_ADJ_IMPORTS, "import { $1, $3 } from $2;\n");
    changeCount++;
  }
  if (changeCount === 0) {
    return code;
  }
  const newLines = importLines.split(/(?<=\n)/).map((line) => formatImport(line, options7));
  return "".concat(...newLines) + codeBody;
}
function formatImport(origLine, options7) {
  let newLines;
  if (!RE_IMPORT_LINE.test(origLine)) {
    return origLine;
  }
  const match = origLine.match(RE_IMPORT_LINE);
  const elements = match[1].split(/, /).sort();
  const target = match[2];
  if (origLine.length > options7.printWidth) {
    const importElements = elements.map((x) => " ".repeat(options7.tabWidth) + x).join(",\n");
    const trailingComma = options7.trailingCommas === "none" ? "" : ",";
    newLines = `import {
${importElements}${trailingComma}
} from ${target};
`;
  } else {
    newLines = `import { ${elements.join(", ")} } from ${target};
`;
  }
  return newLines;
}
var mergeImports, RE_SPLIT_IMPORTS, RE_ADJ_IMPORTS, RE_IMPORT_LINE;
var init_postprocess2 = __esm({
  "src/packages/core-parts/extras/merge-imports/postprocess.ts"() {
    mergeImports = true;
    RE_SPLIT_IMPORTS = /(?<=\nimport \{ [^}]+ \} from [^\n]+\n)/;
    RE_ADJ_IMPORTS = /(?<=\n)import \{ ([^}]+) \} from (\S+);\nimport \{ ([^}]+) \} from \2;\n/;
    RE_IMPORT_LINE = /^import \{ ((?:[^,]+, )+[^,]+) \} from ([^\n]+);\n$/;
  }
});

// src/packages/core-parts/extras/merge-imports/index.ts
var handleParsers2, MergeSimpleImports;
var init_merge_imports = __esm({
  "src/packages/core-parts/extras/merge-imports/index.ts"() {
    init_options2();
    init_postprocess2();
    handleParsers2 = ["babel", "typescript"];
    MergeSimpleImports = {
      options: options2,
      postprocess: (code, opts2) => handleParsers2.includes(opts2.parserName) ? postprocess2(code, opts2) : code
    };
  }
});

// src/packages/core-parts/extras/playwright/options.ts
var options3;
var init_options3 = __esm({
  "src/packages/core-parts/extras/playwright/options.ts"() {
    options3 = {
      playwrightClickGrouping: {
        category: "Format",
        type: "boolean",
        default: false,
        since: "0.6.1",
        description: "Insert a blank line before 'await ...click()'.",
        oppositeDescription: "Do not insert a blank line before 'await ...click()'."
      }
    };
  }
});

// src/packages/core-parts/extras/playwright/postprocess.ts
function setOptions3(options7) {
  pwClickGrouping = !!options7.playwrightClickGrouping;
}
function postprocess3(code, options7) {
  setOptions3(options7);
  if (!pwClickGrouping || !RE_CLICK.test(code)) {
    return code;
  }
  if (pwClickGrouping) {
    code = groupByClickCalls(code);
  }
  return code;
}
function groupByClickCalls(code) {
  if (!pwClickGrouping || !RE_GROUP_CLICK_CALL.test(code)) {
    return code;
  }
  return code.replace(RE_GROUP_CLICK_CALL, "\n$1");
}
var pwClickGrouping, RE_CLICK, RE_GROUP_CLICK_CALL;
var init_postprocess3 = __esm({
  "src/packages/core-parts/extras/playwright/postprocess.ts"() {
    pwClickGrouping = false;
    RE_CLICK = /\.click\(\)/;
    RE_GROUP_CLICK_CALL = /(?<=\{\n|\s*await expect\([^\n]+?(?:\n[ ]*\.[^\n]+?)*;\n)([ ]*await [^\n]+(?:\n\s*\.[^\n]+?)*\s*\.click\([^\n]*\);\n)/g;
  }
});

// src/packages/core-parts/extras/playwright/index.ts
var handleParsers3, Playwright;
var init_playwright = __esm({
  "src/packages/core-parts/extras/playwright/index.ts"() {
    init_options3();
    init_postprocess3();
    handleParsers3 = ["babel", "typescript"];
    Playwright = {
      options: options3,
      postprocess: (code, opts2) => handleParsers3.includes(opts2.parserName) ? postprocess3(code, opts2) : code
    };
  }
});

// src/packages/core-parts/extras/preserve-line/options.ts
var options4;
var init_options4 = __esm({
  "src/packages/core-parts/extras/preserve-line/options.ts"() {
    options4 = {
      preserveFirstBlankLine: {
        category: "Format",
        type: "boolean",
        default: true,
        since: "0.6.1",
        description: "Preserve first blank line of a block.",
        oppositeDescription: "Do not preserve first blank line of a block, Prettier's default behavior."
      },
      preserveLastBlankLine: {
        category: "Format",
        type: "boolean",
        default: true,
        since: "0.6.1",
        description: "Preserve last blank line of a block.",
        oppositeDescription: "Do not preserve last blank line of a block, Prettier's default behavior."
      },
      preserveEolMarker: {
        category: "Format",
        type: "boolean",
        default: true,
        since: "0.6.1",
        description: "Preserve a line if it ends with '//', by applying '// prettier-ignore' to the line.",
        oppositeDescription: "End-of-line '//' is treated as usual by Prettier."
      },
      preserveDotChain: {
        category: "Format",
        type: "boolean",
        default: true,
        since: "0.6.1",
        description: "Preserve line separation of .chain() method calls.",
        oppositeDescription: "Do not preserve line separation of .chain() method calls, Prettier's default behavior."
      }
    };
  }
});

// src/packages/core-parts/extras/preserve-line/align-colon.ts
function alignColon(code, options7) {
  if (!RE_DBL_SLASH_COLON.test(code) || !code || options7.alignObjectProperties === "none") {
    return code;
  }
  const segs = code.split(RE_DBL_SLASH_COLON_SEGMENTS);
  return "".concat(...segs.map((seg) => _alignColonBegin(seg, options7)));
}
function _alignColonBegin(seg, options7) {
  if (!RE_DBL_SLASH_COLON.test(seg) || !seg) {
    return seg;
  }
  const getIndent2 = (line) => {
    const match = line.match(RE_COLON_LINE) || [];
    return match[1] ? match[1].length : 0;
  };
  const lines = seg.split(/(?<=\n)/);
  let alignLines = [];
  const indent = getIndent2(lines[0]);
  while (lines.length && RE_COLON_LINE.test(lines[0]) && getIndent2(lines[0]) === indent) {
    alignLines.push(lines.shift());
  }
  const maxLen = alignLines.map((line = "") => {
    const match = line.match(RE_COLON_LINE) || [];
    return match[2] ? match[2].length : 0;
  }).sort((a, b) => b - a)[0];
  const isAlignColon = options7.alignObjectProperties === "colon";
  alignLines = alignLines.map((line = "") => {
    const match = line.match(RE_COLON_LINE) || [];
    const len = match[2] ? match[2].length : 0;
    return line.replace(
      RE_COLON_LINE,
      isAlignColon ? `$1$2${" ".repeat(maxLen - len)} :` : `$1$2:${" ".repeat(maxLen - len)}`
    );
  });
  return "".concat(...alignLines, ...lines);
}
var RE_DBL_SLASH_COLON, RE_DBL_SLASH_COLON_SEGMENTS, RE_COLON_LINE;
var init_align_colon = __esm({
  "src/packages/core-parts/extras/preserve-line/align-colon.ts"() {
    RE_DBL_SLASH_COLON = /\/\/:\n/;
    RE_DBL_SLASH_COLON_SEGMENTS = /(?<=\n)(?=[^\n]*\/\/:\n)/;
    RE_COLON_LINE = /^(\s*)([^:]+?)\s*:/;
  }
});

// src/packages/core-parts/extras/preserve-line/align-equal.ts
function alignEqual(code) {
  if (!RE_DBL_SLASH_EQUAL.test(code) || !code) {
    return code;
  }
  const segs = code.split(RE_ASSIGN_SEGMENTS);
  return "".concat(...segs.map((seg) => _alignEqualBegin(seg)));
}
function _alignEqualLines(lines) {
  const maxLen = lines.map((line) => line.match(RE_ASSIGN_LINE)[2].length).sort((a, b) => b - a)[0];
  return lines.map((line = "") => {
    const len = line.match(RE_ASSIGN_LINE)[2].length;
    return line.replace(RE_ASSIGN_LINE, `$1$2${" ".repeat(maxLen - len)} $3`);
  });
}
function _alignEqualBegin(seg) {
  const lines = seg.split(/(?<=\n)/);
  if (!seg || !RE_ASSIGN_LINE.test(lines[0])) {
    return seg;
  }
  const doneLines = [];
  const assignLines = [lines.shift()];
  const indent = assignLines[0].match(RE_ASSIGN_LINE)[1].length;
  const reSameIndent = new RegExp(`^\\s{${indent}}\\S`);
  while (lines.length && RE_ASSIGN_LINE.test(lines[0]) && reSameIndent.test(lines[0])) {
    assignLines.push(lines.shift());
  }
  if (assignLines.length > 0) {
    doneLines.push(..._alignEqualLines(assignLines));
  }
  return "".concat(...doneLines, ...lines);
}
var RE_DBL_SLASH_EQUAL, RE_ASSIGN_SEGMENTS, RE_ASSIGN_LINE;
var init_align_equal = __esm({
  "src/packages/core-parts/extras/preserve-line/align-equal.ts"() {
    RE_DBL_SLASH_EQUAL = /\/\/=\n/;
    RE_ASSIGN_SEGMENTS = /(?<=(?:^|\n))(?=[^\n]+\/\/=[ \t]*\n)/;
    RE_ASSIGN_LINE = /^(\s*)(\S[^=]*) ([?&|*/+-]*=\s*\S)/;
  }
});

// src/packages/core-parts/extras/preserve-line/align-slash.ts
function alignTripleSlash(code) {
  if (!RE_TRI_SLASH_SEGS.test(code)) {
    return code;
  }
  const segs = code.split(RE_TRI_SLASH_SEGS);
  return "".concat(...segs.map((seg) => _alignDblSlashLines(seg)));
}
function _alignDblSlashLines(seg) {
  if (!RE_TRI_SLASH_SEGS.test(seg)) {
    return seg;
  }
  const lines = seg.split(/(?<=\n)/);
  let alignLines = [];
  while (lines.length && lines[0].match(/\S/)) {
    alignLines.push(lines.shift());
  }
  const maxLen = alignLines.map((line = "") => {
    const match = line.match(RE_DBL_SLASH_LINE) || [];
    return match[1] ? match[1].length : 0;
  }).sort((a, b) => b - a)[0];
  alignLines = alignLines.map((line = "") => {
    const match = line.match(RE_DBL_SLASH_LINE) || [];
    const len = match[1] ? match[1].length : 0;
    return line.replace(RE_DBL_SLASH_LINE, `$1${" ".repeat(maxLen - len)}  //`);
  });
  return "".concat(...alignLines, ...lines);
}
var RE_TRI_SLASH_SEGS, RE_DBL_SLASH_LINE;
var init_align_slash = __esm({
  "src/packages/core-parts/extras/preserve-line/align-slash.ts"() {
    RE_TRI_SLASH_SEGS = /(?<=\n|^)(?=[^\n]*\/\/\/[=:]?\n)/;
    RE_DBL_SLASH_LINE = /^(?:(.*?)\s+|)\/\//;
  }
});

// src/packages/core-parts/extras/preserve-line/parser.ts
function preprocess(code, options7) {
  if (options7.preserveFirstBlankLine) {
    code = code.replace(RE_OPEN_BLANK_LINE, `$1${BR}
`);
  }
  if (options7.preserveLastBlankLine) {
    code = code.replace(RE_CLOSE_BLANK_LINE, `${BR}
$1`);
  }
  if (options7.preserveEolMarker) {
    code = code.replace(RE_DOUBLE_SLASH_EOL, `${IGNORE}
$1`);
    code = code.replace(RE_DBL_SLASH_PRESERVE_SPACE, " //$1");
  }
  if (options7.preserveDotChain) {
    code = code.replace(RE_DOT_CHAIN_METHOD, `${DOT}
$1`);
    while (code.match(RE_DOT_NO_PRESERVE)) {
      code = code.replace(RE_DOT_NO_PRESERVE, "$1").replace(RE_DOT_NO_PRESERVE, "$1");
    }
  }
  return code;
}
function postprocess4(code, options7) {
  if (options7.preserveFirstBlankLine || options7.preserveLastBlankLine) {
    code = code.replace(RE_BR_LINE, "");
  }
  if (options7.preserveDotChain) {
    while (code.match(RE_CONCAT_DOTS)) {
      code = code.replace(RE_CONCAT_DOTS, "$1$2").replace(RE_CONCAT_DOTS, "$1$2");
    }
    while (code.match(RE_CONCAT_DOTS_BEFORE)) {
      code = code.replace(RE_CONCAT_DOTS_BEFORE, "$1$2").replace(RE_CONCAT_DOTS_BEFORE, "$1$2");
    }
    code = code.replace(RE_DOT_LINE, "");
  }
  if (options7.preserveEolMarker) {
    code = code.replace(RE_DBL_SLASH_IGNORE, "");
    code = code.replace(RE_DBL_RESTORE_SPACE, " $1");
    code = alignEqual(code);
    code = alignColon(code, options7);
    code = alignTripleSlash(code);
  }
  return code;
}
var RE_OPEN_BLANK_LINE, BR, RE_BR_LINE, RE_CLOSE_BLANK_LINE, IGNORE, RE_DOUBLE_SLASH_EOL, RE_DBL_SLASH_PRESERVE_SPACE, RE_DBL_SLASH_IGNORE, RE_DBL_RESTORE_SPACE, RE_DOT_CHAIN_METHOD, DOT, RE_DOT_LINE, NO_PRESERVE, RE_DOT_NO_PRESERVE, RE_CONCAT_DOTS, RE_CONCAT_DOTS_BEFORE;
var init_parser = __esm({
  "src/packages/core-parts/extras/preserve-line/parser.ts"() {
    init_align_colon();
    init_align_equal();
    init_align_slash();
    RE_OPEN_BLANK_LINE = /([{[(=:|&]\s*\n)(?:\s*\n)+/g;
    BR = "//__BR__";
    RE_BR_LINE = new RegExp(`[ ]*${BR}$`, "mg");
    RE_CLOSE_BLANK_LINE = /(?<=\n)(?:\s*\n)+(\s*[}\])])/g;
    IGNORE = "// prettier-ignore";
    RE_DOUBLE_SLASH_EOL = /^(.*\/\/)$/gm;
    RE_DBL_SLASH_PRESERVE_SPACE = /(?<=\S)[ ]([ ]+\/\/(?:[^\n]+\/\/)?(?:=|:|\/|\/=)?)$/gm;
    RE_DBL_SLASH_IGNORE = new RegExp(`[ \\t]*${IGNORE}\\n(?=[^\\n]*//\\n)`, "g");
    RE_DBL_RESTORE_SPACE = /(?<=\S)[ ]\/\/(\s+\/\/(?:[^\n]+\/\/)?(?:=|:|\/|\/=)?)$/gm;
    RE_DOT_CHAIN_METHOD = /^(\s*\.[a-z_]\w*[(.])/gim;
    DOT = "//__DOT__";
    RE_DOT_LINE = new RegExp(`[ \\t]*${DOT}\\n`, "g");
    NO_PRESERVE = "// no-preserve";
    RE_DOT_NO_PRESERVE = new RegExp(`(${NO_PRESERVE}\\n(?:[ \\t]*[^\\n]+\\n)+?)${DOT}\\n`, "g");
    RE_CONCAT_DOTS = new RegExp(
      `(${DOT}\\n[ \\t]*\\.[^\\n]+(?:\\n[^\\n]+)*?(?:\\n[^\\n]*)?\\))\\n[ \\t]*(\\.[^\\n]+)`,
      "g"
    );
    RE_CONCAT_DOTS_BEFORE = new RegExp(
      `(?<=^|\\n)([ \\t]*[^\\./ \\t][^\\n]+)\\n[ \\t]*(\\.[^\\n]+\\n)(?=(?:[ \\t]*\\.[^\\n]+\\n)*[ \\t]*${DOT})`
    );
  }
});

// src/packages/core-parts/extras/preserve-line/index.ts
var handleParsers4, PreserveLine;
var init_preserve_line = __esm({
  "src/packages/core-parts/extras/preserve-line/index.ts"() {
    init_options4();
    init_parser();
    handleParsers4 = ["babel", "typescript"];
    PreserveLine = {
      options: options4,
      preprocess: (code, opts2) => handleParsers4.includes(opts2.parserName) ? preprocess(code, opts2) : code,
      postprocess: (code, opts2) => handleParsers4.includes(opts2.parserName) ? postprocess4(code, opts2) : code
    };
  }
});

// src/packages/core-parts/extras/return-parentheses/options.ts
var options5;
var init_options5 = __esm({
  "src/packages/core-parts/extras/return-parentheses/options.ts"() {
    options5 = {
      returnParentheses: {
        category: "Format",
        type: "boolean",
        default: false,
        since: "0.6.1",
        description: "Don't wrap multi-line 'return' values in parentheses.",
        oppositeDescription: "Wrap multi-line 'return' values in parentheses."
      }
    };
  }
});

// src/packages/core-parts/extras/return-parentheses/strip-return-paren.ts
function stripReturnParentheses(code, options7) {
  let lines = code.split(/(?<=\n)/);
  const linesDone = [];
  let match;
  if (options7) {
    semi = options7?.semi ?? semi;
  }
  while (lines.length > 0) {
    while (lines.length > 0 && !(match = lines[0].match(RE_RETURN_OR_ARROW_FN_PAREN))) {
      linesDone.push(lines.shift());
    }
    if (lines.length === 0) {
      break;
    }
    const indent = match?.groups?.INDENT;
    const indentMoreRE = new RegExp(`^(?:${indent}[ \\t]+\\S|${indent}[ \\t]+\\n)`);
    const endReturnRE = new RegExp(`^${indent}\\);?
`);
    const returnBlock = [lines.shift()];
    while (lines.length > 0 && (indentMoreRE.test(lines[0]) || endReturnRE.test(lines[0]))) {
      returnBlock.push(lines.shift());
      const endReturn = endReturnRE.test(returnBlock[returnBlock.length - 1]);
      if (endReturn) {
        break;
      }
    }
    if (endReturnRE.test(returnBlock[returnBlock.length - 1])) {
      linesDone.push(stripReturnOuterParentheses(returnBlock));
    } else {
      linesDone.push(returnBlock.shift());
      lines = [...returnBlock, ...lines];
    }
  }
  return linesDone.join("");
}
function stripReturnOuterParentheses(returnBlock) {
  if (returnBlock.length <= 2) {
    return returnBlock.join("");
  }
  const match1 = returnBlock[0].match(RE_INDENT);
  const indent1 = match1?.groups?.INDENT;
  const match2 = returnBlock[1].match(RE_INDENT);
  const indent2 = match2?.groups?.INDENT;
  const innerFirstLastLine = returnBlock[1] + returnBlock[returnBlock.length - 2].replace(/^\s+/, "");
  const shouldDedent = innerFirstLastLine.match(/\(\n\)|\{\n\}|\[\n\]|>\n</) || returnBlock[2].match(/^\s*\./);
  const firstLine = returnBlock.splice(0, 2).join("").replace(RE_RETURN_OR_ARROW_FN_PAREN, "$1$2 ");
  returnBlock.splice(returnBlock.length - 1, 1);
  if (semi) {
    returnBlock.push((returnBlock.pop() || "").replace(/(?<!;)\n$/, ";\n"));
  }
  if (shouldDedent) {
    returnBlock = returnBlock.map((line) => line.replace(new RegExp(`^${indent2}`), `${indent1}`));
  }
  return stripReturnParentheses([firstLine, ...returnBlock].join(""));
}
var RE_RETURN_OR_ARROW_FN_PAREN, semi, RE_INDENT;
var init_strip_return_paren = __esm({
  "src/packages/core-parts/extras/return-parentheses/strip-return-paren.ts"() {
    RE_RETURN_OR_ARROW_FN_PAREN = // @ts-ignore
    /^(?<INDENT>[ \t]*)(?<RETURN_OR_ARROW_FN>return|[a-zA-Z_][^\n]+\s*=>)[ \t]*\([ \t]*\n[ \t]*/;
    semi = true;
    RE_INDENT = /^(?<INDENT>[ \t]*)/;
  }
});

// src/packages/core-parts/extras/return-parentheses/postprocess.ts
function setOptions4(options7) {
  returnParentheses = options7.returnParentheses ?? returnParentheses;
}
function postprocess5(code, options7) {
  setOptions4(options7);
  const detectReturnParenthesis = RE_RETURN_OR_ARROW_FN_PAREN2.test(code);
  if (!returnParentheses && detectReturnParenthesis) {
    return stripReturnParentheses(code, options7);
  }
  return code;
}
var returnParentheses, RE_RETURN_OR_ARROW_FN_PAREN2;
var init_postprocess4 = __esm({
  "src/packages/core-parts/extras/return-parentheses/postprocess.ts"() {
    init_strip_return_paren();
    returnParentheses = false;
    RE_RETURN_OR_ARROW_FN_PAREN2 = /(?:^|(?<=\n))\s*(?:return|[a-zA-Z_][^\n]+\s*=>)\s+\(\s*\n/;
  }
});

// src/packages/core-parts/extras/return-parentheses/index.ts
var handleParsers5, ReturnParentheses;
var init_return_parentheses = __esm({
  "src/packages/core-parts/extras/return-parentheses/index.ts"() {
    init_options5();
    init_postprocess4();
    handleParsers5 = ["babel", "typescript", "angular"];
    ReturnParentheses = {
      options: options5,
      postprocess: (code, opts2) => handleParsers5.includes(opts2.parserName) ? postprocess5(code, opts2) : code
    };
  }
});

// src/packages/core-parts/extras/index.ts
var Extras;
var init_extras = __esm({
  "src/packages/core-parts/extras/index.ts"() {
    init_align_obj_props2();
    init_merge_imports();
    init_playwright();
    init_preserve_line();
    init_return_parentheses();
    Extras = {
      options: {
        ...PreserveLine.options,
        ...AlignObjectProperties.options,
        ...MergeSimpleImports.options,
        ...Playwright.options,
        ...ReturnParentheses.options
      },
      preprocess: (code, options7) => {
        code = PreserveLine.preprocess(code, options7);
        return code;
      },
      postprocess: (code, options7) => {
        code = PreserveLine.postprocess(code, options7);
        code = MergeSimpleImports.postprocess(code, options7);
        code = AlignObjectProperties.postprocess(code, options7);
        code = Playwright.postprocess(code, options7);
        code = ReturnParentheses.postprocess(code, options7);
        return code;
      }
    };
  }
});

// src/packages/core-parts/options.ts
var options6;
var init_options6 = __esm({
  "src/packages/core-parts/options.ts"() {
    init_extras();
    options6 = Extras.options;
  }
});

// src/packages/core-parts/index.ts
function makePatches(oldStr, newStr) {
  if (oldStr === newStr) {
    return [];
  }
  const patches = [];
  let temporaryText = "";
  let mode = "even" /* EVEN */;
  diffLines(oldStr, newStr).filter((change) => "added" in change && "removed" in change).forEach((change) => {
    if (!change.added && change.removed) {
      if (mode === "even" /* EVEN */) {
        temporaryText = change.value;
        mode = "odd" /* ODD */;
      } else {
        patches.push({ from: temporaryText, to: "" });
        temporaryText = change.value;
      }
    } else if (change.added && !change.removed) {
      if (mode === "even" /* EVEN */) {
        patches.push({ from: "", to: change.value });
      } else {
        patches.push({ from: temporaryText, to: change.value });
        mode = "even" /* EVEN */;
      }
    }
  });
  return patches;
}
function applyPatches(text, patches) {
  return patches.reduce((patchedPrevText, { from, to }) => patchedPrevText.replace(from, to), text);
}
var init_core_parts = __esm({
  "src/packages/core-parts/index.ts"() {
    init_lib();
    init_extras();
    init_options6();
  }
});

// src/packages/v2-plugin/parsers.ts
function sequentialFormattingAndTryMerging(options7, plugins, languageImplementedPlugin) {
  const customLanguageSupportedPlugins = languageImplementedPlugin ? [languageImplementedPlugin] : [];
  const { originalText } = options7;
  const sequentialFormattingOptions = {
    ...options7,
    rangeEnd: Infinity,
    endOfLine: "lf",
    plugins: customLanguageSupportedPlugins
  };
  const firstFormattedText = (0, import_prettier.format)(originalText, sequentialFormattingOptions);
  const patches = [];
  const sequentiallyMergedText = plugins.reduce((formattedPrevText, plugin) => {
    const temporaryFormattedText = (0, import_prettier.format)(formattedPrevText, {
      ...sequentialFormattingOptions,
      plugins: [...customLanguageSupportedPlugins, plugin]
    });
    const temporaryFormattedTextWithoutPlugin = (0, import_prettier.format)(
      temporaryFormattedText,
      sequentialFormattingOptions
    );
    patches.push(...makePatches(temporaryFormattedTextWithoutPlugin, temporaryFormattedText));
    if (patches.length === 0) {
      return temporaryFormattedText;
    }
    return applyPatches(temporaryFormattedTextWithoutPlugin, patches);
  }, firstFormattedText);
  return sequentiallyMergedText;
}
function transformParser(parserName, defaultParser, languageName) {
  return {
    ...defaultParser,
    parse: (text, parsers3, options7) => {
      const plugins = options7.plugins.filter((plugin) => typeof plugin !== "string");
      const pluginIndex = plugins.findIndex(
        (plugin) => Object.values(plugin.parsers ?? {}).every(
          (parser) => parser.astFormat === "merging-ast"
        ) && Object.entries(plugin.printers ?? {}).every(
          ([astFormat, printer]) => astFormat === "merging-ast" && Object.keys(printer).every((key) => key === "print")
        )
      );
      if (pluginIndex === -1) {
        throw new Error(
          "The structure of this plugin may have changed. If it's not in development, you may need to report an issue."
        );
      }
      let languageImplementedPlugin;
      if (languageName) {
        languageImplementedPlugin = plugins.slice(0, pluginIndex).filter((plugin) => plugin.languages?.some((language) => language.name === languageName)).at(0);
        if (!languageImplementedPlugin) {
          throw new Error(
            `There doesn't seem to be any plugin that supports ${languageName} formatting.`
          );
        }
      }
      const optionsExtras = { ...options7, parserName };
      let code = Extras.preprocess(text, optionsExtras);
      const parserImplementedPlugins = plugins.slice(0, pluginIndex).filter((plugin) => plugin.parsers?.[parserName]);
      code = sequentialFormattingAndTryMerging(
        {
          ...options7,
          originalText: code
        },
        parserImplementedPlugins,
        languageImplementedPlugin
      );
      code = Extras.postprocess(code, optionsExtras);
      return {
        type: "FormattedText",
        body: code
      };
    },
    astFormat: "merging-ast"
  };
}
var import_prettier, import_parser_babel, import_parser_html, import_parser_typescript, parsers;
var init_parsers = __esm({
  "src/packages/v2-plugin/parsers.ts"() {
    "use strict";
    init_core_parts();
    import_prettier = require("prettier");
    import_parser_babel = require("prettier/parser-babel");
    import_parser_html = require("prettier/parser-html");
    import_parser_typescript = require("prettier/parser-typescript");
    parsers = {
      babel: transformParser("babel", import_parser_babel.parsers.babel),
      typescript: transformParser("typescript", import_parser_typescript.parsers.typescript),
      angular: transformParser("angular", import_parser_html.parsers.angular),
      html: transformParser("html", import_parser_html.parsers.html),
      vue: transformParser("vue", import_parser_html.parsers.vue),
      astro: transformParser("astro", {}, "astro"),
      svelte: transformParser("svelte", {}, "svelte")
    };
  }
});

// src/packages/v2-plugin/printers.ts
function createPrinter() {
  function main(path, options7, print) {
    const node = path.getValue();
    if (node.type === "FormattedText") {
      return node.body;
    }
    throw new Error(`Unknown node type: ${node?.type}`);
  }
  return {
    print: main
  };
}
var printers;
var init_printers = __esm({
  "src/packages/v2-plugin/printers.ts"() {
    "use strict";
    printers = {
      "merging-ast": createPrinter()
    };
  }
});

// src/packages/v2-plugin/index.ts
var v2_plugin_exports = {};
__export(v2_plugin_exports, {
  options: () => options6,
  parsers: () => parsers,
  printers: () => printers
});
var init_v2_plugin = __esm({
  "src/packages/v2-plugin/index.ts"() {
    "use strict";
    init_core_parts();
    init_parsers();
    init_printers();
  }
});

// src/packages/v3-plugin/parsers.ts
async function sequentialFormattingAndTryMerging2(options7, plugins, languageImplementedPlugin) {
  const customLanguageSupportedPlugins = languageImplementedPlugin ? [languageImplementedPlugin] : [];
  const { originalText } = options7;
  const sequentialFormattingOptions = {
    ...options7,
    rangeEnd: Infinity,
    endOfLine: "lf",
    plugins: customLanguageSupportedPlugins
  };
  const firstFormattedTextPromise = (0, import_prettier2.format)(originalText, sequentialFormattingOptions);
  const patches = [];
  const sequentiallyMergedText = await plugins.reduce(
    async (formattedPrevTextPromise, plugin) => {
      const formattedPrevText = await formattedPrevTextPromise;
      const temporaryFormattedText = await (0, import_prettier2.format)(formattedPrevText, {
        ...sequentialFormattingOptions,
        plugins: [...customLanguageSupportedPlugins, plugin]
      });
      const temporaryFormattedTextWithoutPlugin = await (0, import_prettier2.format)(
        temporaryFormattedText,
        sequentialFormattingOptions
      );
      patches.push(...makePatches(temporaryFormattedTextWithoutPlugin, temporaryFormattedText));
      if (patches.length === 0) {
        return temporaryFormattedText;
      }
      return applyPatches(temporaryFormattedTextWithoutPlugin, patches);
    },
    firstFormattedTextPromise
  );
  return sequentiallyMergedText;
}
function transformParser2(parserName, defaultParser, languageName) {
  return {
    ...defaultParser,
    parse: async (text, options7) => {
      const plugins = options7.plugins.filter((plugin) => typeof plugin !== "string");
      const pluginIndex = plugins.findIndex(
        (plugin) => Object.values(plugin.parsers ?? {}).every(
          (parser) => parser.astFormat === "merging-ast"
        ) && Object.entries(plugin.printers ?? {}).every(
          ([astFormat, printer]) => astFormat === "merging-ast" && Object.keys(printer).every((key) => key === "print")
        )
      );
      if (pluginIndex === -1) {
        throw new Error(
          "The structure of this plugin may have changed. If it's not in development, you may need to report an issue."
        );
      }
      let languageImplementedPlugin;
      if (languageName) {
        languageImplementedPlugin = plugins.slice(0, pluginIndex).filter((plugin) => plugin.languages?.some((language) => language.name === languageName)).at(0);
        if (!languageImplementedPlugin) {
          throw new Error(
            `There doesn't seem to be any plugin that supports ${languageName} formatting.`
          );
        }
      }
      const optionsExtras = { ...options7, parserName };
      let code = Extras.preprocess(text, optionsExtras);
      const parserImplementedPlugins = plugins.slice(0, pluginIndex).filter((plugin) => plugin.parsers?.[parserName]);
      code = await sequentialFormattingAndTryMerging2(
        {
          ...options7,
          originalText: code
        },
        parserImplementedPlugins,
        languageImplementedPlugin
      );
      code = Extras.postprocess(code, optionsExtras);
      return {
        type: "FormattedText",
        body: code
      };
    },
    astFormat: "merging-ast"
  };
}
var import_prettier2, import_babel, import_html, import_typescript, parsers2;
var init_parsers2 = __esm({
  "src/packages/v3-plugin/parsers.ts"() {
    "use strict";
    init_core_parts();
    import_prettier2 = require("prettier");
    import_babel = require("prettier/plugins/babel");
    import_html = require("prettier/plugins/html");
    import_typescript = require("prettier/plugins/typescript");
    parsers2 = {
      babel: transformParser2("babel", import_babel.parsers.babel),
      typescript: transformParser2("typescript", import_typescript.parsers.typescript),
      angular: transformParser2("angular", import_html.parsers.angular),
      html: transformParser2("html", import_html.parsers.html),
      vue: transformParser2("vue", import_html.parsers.vue),
      astro: transformParser2("astro", {}, "astro"),
      svelte: transformParser2("svelte", {}, "svelte")
    };
  }
});

// src/packages/v3-plugin/printers.ts
function createPrinter2() {
  function main(path, options7, print) {
    const { node } = path;
    if (node.type === "FormattedText") {
      return node.body;
    }
    throw new Error(`Unknown node type: ${node?.type}`);
  }
  return {
    print: main
  };
}
var printers2;
var init_printers2 = __esm({
  "src/packages/v3-plugin/printers.ts"() {
    "use strict";
    printers2 = {
      "merging-ast": createPrinter2()
    };
  }
});

// src/packages/v3-plugin/index.ts
var v3_plugin_exports = {};
__export(v3_plugin_exports, {
  options: () => options6,
  parsers: () => parsers2,
  printers: () => printers2
});
var init_v3_plugin = __esm({
  "src/packages/v3-plugin/index.ts"() {
    "use strict";
    init_core_parts();
    init_parsers2();
    init_printers2();
  }
});

// src/index.ts
var import_prettier3 = __toESM(require("prettier"));
if (import_prettier3.default.version.startsWith("2.")) {
  module.exports = (init_v2_plugin(), __toCommonJS(v2_plugin_exports));
} else {
  module.exports = (init_v3_plugin(), __toCommonJS(v3_plugin_exports));
}
